### DISCUSSION TITLE
How should a response with content "application/json" be handled where json.Marshal() returns an error?

### DISCUSSION THREAD

**Discussion by nchelluri** [oapi-codegen/oapi-codegen#1674] [Q&A] [answered]
Suppose we have an API response with `content` `application/json` following some `schema`, but the response cannot be run through `json.Marshal()` without that call returning an error. An example:

```go
type someStruct struct {
	Val float64
}
```

The following value for that struct will not marshal:

```go
someStruct{Val: zero / zero}
```

[Go Playground link](https://go.dev/play/p/pzIrwpX2_Kk)

Now suppose we have a server handler func with a signature like `func (s *Server) GetSomeStruct(ctx context.Context, request api.GetSomeStructRequestObject) (api.GetSomeStructResponseObject, error)`.

If we try to return such a value in our handler func, like by calling `return api.GetSomeStruct200JSONResponse(s), nil` then we will get an HTTP response with a 200 status, but a body of `json: unsupported value: NaN`, which is not valid JSON. I am wondering how to handle this case. I would like to have control over such cases. Thank you.

**Comment by mromaszewicz** [comment, marked as answer]
Yeah, good question. We don't solve that problem here, because, fundamentally, you control the content of the object that you marshal. In this specific case, you'd have to check for NaN before marshalling and handle that - either zero it out or return a different kind of result code if it's an actual error.

I have hundreds of API endpoints in production using oapi-codegen, and some of them do indeed produce data that can't be marshalled, so in the places that this happens, I validate before the marshaling. A common case for me is also consistency. A `nil` slice marshals as `Foo: nil` but I want `Foo: []`, so I special case all of these too.

The only automatic way would be a custom JSON marshaler that can deal with degenerate cases.

**Reply by nchelluri** [reply]
I understand now; thanks for taking the time to explain.
